contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -742,51 +768,152 @@ have a built-in `nil` value.
+
+Static default values are specific to each API version.  The default field
+values applied when creating an object with the "v1" API may be different than
+the values applied when using the "v2" API.  In most cases, these values are
+defined as literal values by the API version (e.g. "if this field is not
+specified it defaults to 0").  In some cases, these values may be
+deterministically derived from other fields (e.g. "if this field is not
+specified it defaults to the value of otherField").  In very rare cases, these
+values may be allocated from some pool or determined by some other method (e.g.
+Service's IP and IP-family related fields).
+
+These values are applied synchronously by the API server during the conversion
+from API-versioned declarative configuration to internal objects representing
+the desired state (`Spec`) of the resource. When the API call completes, all
+static defaults will have been set.  Subsequent GETs of the resource will
+include the default values explicitly.
+
+Static defaults are the best choice for values which are logically required,
+but which have a value that works well for most users.  Static defaulting
+must not consider any state except the object being operated upon (and the
+complexity of Service API stands as an example of why).
 
 Default values can be specified on a field using the `+default=` tag. Primitives
 will have their values directly assigned while structs will go through the
 JSON unmarshalling process. Fields that do not have an `omitempty` json tag will
 default to the zero value of their corresponding type if no default is assigned.
 
-Refer to [defaulting docs](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting) for more information.
+Refer to [defaulting docs](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting)
+for more information.
+
+### Admission Controlled Defaults
+
+In some cases, it is useful to set a default value which is not derived from
+the object in question.  For example, when creating a PersistentVolumeClaim,
+the storage class must be specified.  For many users, the best answer is
+"whatever the cluster admin has decided for the default".  StorageClass is a
+different API than PersistentVolumeClaim, and which one is denoted as the
+default may change at any time.  Thus this is not eligible for static
+defaulting.
 
-API version-specific default values are set by the API server.
+Instead, we can provide a built-in admission controller or a
+MutatingWebhookConfiguration.  Unlike static defaults, these may consider
+external state (such as annotations on StorageClass objects) when deciding
+default values.  These admission controllers are strictly optional and can be
+disabled.  As such, fields which are initialized this way must be strictly
+optional.
 
-## Late Initialization
+Like static defaults, these are run synchronously to the API operation in
```

Comment:
do we want to add a mention that this defaulting runs "before" static defaulting? and may override it?

Reply:
there is some text that get drop about this

> Although the apiserver Admission Control stage acts prior to object creation,
Admission Control plugins should follow the Late Initialization conventions
too, to allow their implementation to be later moved to a 'controller', or to
client libraries.

Reply:
Static defaulting runs before admission.

* User sends versioned object to API server
* API server decodes, applies versioned defaults
* Mutating admission is invoked
  * Compiled-in admission operates on internal objects
  * Admission webhooks return patches which are applied to a versioned object (and then static defaulting is applied again, actually)

Reply:
oh, showing my ignorance :smile:

follow up question, how admission can know if the field was really defaulted by the static default was the user setting it to the same value of the default?

Reply:
> how admission can know if the field was really defaulted by the static default was the user setting it to the same value of the default?

good question, it can't :-/

Reply:
Yeah, this is why Service defaults IPFamilyPolicy in the registry stack, so we CAN tell.  Realistically, we should carry information about values that were defaulted vs user-provided, but we do not track that anywhere.